Arm/Memory/AddressNormalization.lean

```diff
@@ -0,0 +1,267 @@
+  If we thus normalize our expressions to have constants on the left,
+  and if we constFoldiate our additions to be to the left, we will naturally perform
+  constant folding:
+
+  (a) (x + c) -> (c + x).
+  (b) x + (y + z) -> (x + y) + c.
+
+  Observe an example:
+
+  ((x + 10) + 20)
+  -b-> (20 + (x + 10))
+  -b-> (20 + (10 + x))
+  -a-> (20 + 10) + x
+  -reduceAdd-> 30 + x
+
+2. Canonicalizing (a + b) % n → a % n + b % n by exploiting `bv_omega`,
+   and eventually, `simp_mem`.
+-/
+import Lean
+import Arm.Memory.Attr
+import Arm.Attr
+
+open Lean Meta Elab Simp
+
+
+theorem Nat.mod_eq_sub {x y : Nat} (h : x ≥ y) (h' : x - y < y) :
+    x % y = x - y := by
+  rw [Nat.mod_eq_sub_mod h, Nat.mod_eq_of_lt h']
+
+private def mkLTNat (x y : Expr) : Expr :=
+  mkAppN (.const ``LT.lt [levelZero]) #[mkConst ``Nat, mkConst ``instLTNat, x, y]
+
+private def mkLENat (x y : Expr) : Expr :=
+  mkAppN (.const ``LE.le [levelZero]) #[mkConst ``Nat, mkConst ``instLENat, x, y]
+
+private def mkGENat (x y : Expr) : Expr := mkLENat y x
+
+private def mkSubNat (x y : Expr) : Expr :=
+  let lz := levelZero
+  let nat := mkConst ``Nat
+  let instSub := mkConst ``instSubNat
+  let instHSub := mkAppN (mkConst ``instHSub [lz]) #[nat, instSub]
+  mkAppN (mkConst ``HSub.hSub [lz, lz, lz]) #[nat, nat, nat, instHSub, x, y]
+
+/--
+Given an expression of the form `n#w`, return the value of `n` if it is a ground constant.
+-/
+def getBitVecOfNatValue? (e : Expr) : (Option (Expr × Expr)) :=
+  match_expr e with
+  | BitVec.ofNat nExpr vExpr => some (nExpr, vExpr)
+  | _ => none
+
+/-- Try to build a proof for `ty` by reduction to `bv_omega`. -/
+@[inline] def proveByBvOmega (ty : Expr) : SimpM Step := do
+  let proof : Expr ← mkFreshExprMVar ty
+  let g := proof.mvarId!
+  let some g ← g.falseOrByContra
+    | return .continue
+  try
+    g.withContext (do Lean.Elab.Tactic.Omega.omega (← getLocalHyps).toList g {})
+  catch _ =>
+    return .continue
+  return .done { expr := ty, proof? := proof }
+
+/-- Try to build a proof for `ty` by reduction to `omega`. -/
+@[inline] def proveByOmega (ty : Expr) : SimpM Step := do
+  let proof : Expr ← mkFreshExprMVar ty
+  let g := proof.mvarId!
+  let some g ← g.falseOrByContra
+    | return .continue
+  try
+    g.withContext (do Lean.Elab.Tactic.Omega.omega (← getLocalHyps).toList g {})
+  catch _ =>
+    return .continue
+  return .done { expr := ty, proof? := proof }
+
+def processingEmoji : String := "⚙️"
+
+-- x % n = x if x < n
+@[inline] def reduceModOfLt (x : Expr) (n : Expr) : SimpM Step := do
+  trace[Tactic.address_normalization] "{processingEmoji} reduceModOfLt '{x} % {n}'"
+  let ltTy := mkLTNat x n
+  let Step.done { expr := _, proof? := some p} ← proveByOmega ltTy
+    | return .continue
+  let eqProof ← mkAppM ``Nat.mod_eq_of_lt #[p]
+  trace[Tactic.address_normalization] "{checkEmoji} reduceModOfLt '{x} % {n}'"
+  return .done { expr := x, proof? := eqProof : Result }
```

Comment:
I'm really looking forward to a version of simp that can take auto-tactics as dischargers, that would remove the need for at least a few of these simprocs.

In the short term, yes, let's add them to LNSym, but for the upstreaming story, could we develop something slightly more general which takes in a simp-lemma with auto-tactics, and creates a simproc that does the right thing?

Reply:
@alexkeizer that makes a ton of sense.